audit: improve ssl/tls detection

DomT4 opened the pull request:
I don’t know how maintainers are going to feel about this, to be honest. If it’s too clunky, perhaps we could externalise the entire two main blocks here and then require that file into the audit instead?

Basically, I’m pushing changes here to better detect a wide-array of SSL/TLS available links that either have no auto-redirect in place or is a common linking error in formulae. I haven’t spotted any false positives yet, but obviously, feel free to try and break the changes and I’ll fix as necessary ;).

IMO, this would allow us gradual updates without having to mass-update everything at once and stress the bot and inform users they have hundreds of updates pending when really it’s just style/basic changes. It's also generally a decent security step, even if a basic one.

DomT4 commented:
@jacknagel, @mikemcquaid, @tdsmith, @adamv, @mistydemeo may want to review, given I'm fiddling with Brew's guts here.

DomT4 commented:
Some sample readout:

![screen shot 2015-01-04 at 23 48 45](https://cloud.githubusercontent.com/assets/6998367/5608050/4855bae8-946c-11e4-835e-124cc238dd9b.png)

vszakats commented:
Thumbs up!

You may want to consider mandating ssl/tls access 
for git/hg repositories hosted on:

gitorious.org, github.com, bitbucket.org

[ 'git:' protocol is unsecure: http://git-scm.com/book/ch4-1.html#The-Git-Protocol ]

vszakats commented:
Not terribly widely used, but `s3.amazonaws.com` may also always use ssl/tls. Best to avoid putting the bucket as a subdomain, as names with dots will make the certificate check (and thus https) fail.
This form will always work: `https://s3.amazonaws.com/my-bucket-name/path/file`

DomT4 commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -218,6 +218,27 @@ def audit_urls
       problem "Google Code homepage should end with a slash (URL is #{homepage})."
     end
 
+    # Automatic re-direct exists, but this is another hugely common error.
```
There's an unnecessary ` - ` in this sentence. I'll kill it when I incorporate whatever feedback maintainers give later.

DomT4 commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -225,9 +246,35 @@ def audit_urls
+      # Skip the main url link, as it can't be made SSL/TLS yet.
+      next if p =~ %r[/ftpmirror\.gnu\.org]
+
+      # No auto-direct as of Jan 2015.
```
Likewise, all of these should read ` No auto-redirect `. Will fix post-feedback.

jacknagel commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -225,9 +246,35 @@ def audit_urls
+      # No auto-direct as of Jan 2015.
+      if p =~ %r[^http://mirrors\.kernel\.org/]
+        problem "mirrors.kernel links should be https://, not http (url is #{p})."
+      end
```
since all of these are mutually exclusive, it can be written as

```rb
case p
when %r[pattern]
  problem "..."
when %r[another pattern]
  problem "..."
end
```

DomT4 replied:
```
case p
when %r[^http://mirrors\.kernel\.org/]
   problem "mirrors.kernel links should be https://, not http (url is #{p})."
when %r[^http://fossies\.org]
   problem "Fossies links should be https://, not http (url is #{p})."
end
```

Like that?

jacknagel replied:
yep.

DomT4 replied:
Neat. Should help make the block look a little tighter. Thanks!

Sent from my iPhone

> On 5 Jan 2015, at 04:43, Jack Nagel <notifications@github.com> wrote:
> 
> In Library/Homebrew/cmd/audit.rb:
> 
> > +      end
> > +
> > +      # This does redirect, but it's an incredibly common linking mistake.
> > +      if p =~ %r[^http://code\.google\.com]
> > +        problem "code.google.com links should be https://, not http (url is #{p})."
> > +      end
> > +
> > +      # No auto-direct as of Jan 2015.
> > +      if p =~ %r[^http://fossies\.org]
> > +        problem "Fossies links should be https://, not http (url is #{p})."
> > +      end
> > +
> > +      # No auto-direct as of Jan 2015.
> > +      if p =~ %r[^http://mirrors\.kernel\.org/]
> > +        problem "mirrors.kernel links should be https://, not http (url is #{p})."
> > +      end
> yep.
> 
> —
> Reply to this email directly or view it on GitHub.
>

DomT4 commented:
> You may want to consider mandating ssl/tls access
for git/hg repositories hosted on: gitorious.org, github.com, bitbucket.org

We already do Github: 

```
    # Check for git:// GitHub repo urls, https:// is preferred.
    urls.grep(%r[^git://[^/]*github\.com/]) do |u|
      problem "Use https:// URLs for accessing GitHub repositories (url is #{u})."
    end
```

I've added one for gitorious as well. Not sure about BitBucket, right now, it looks like there are zero formulae using ` git://bitbucket `, and given most formula authors reference other formulas when creating their own, it might not catch on. Unsure. If the maintainers want a BitBucket rule as well, I'll add one.

Pushed the style changes I referenced earlier, and pushed the considerably better formatting Jack suggested as well.

mikemcquaid commented:
This works for me, any objections @Homebrew/owners?

vszakats commented:
some more domains that support ssl/tls:

bugs.debian.org
*.ietf.org
dl.google.com
*.googlecode.com
s3.amazonaws.com (mentioned above) 
*.s3.amazonaws.com
*.github.io (direct file downloads)
*.python.org
dl.dropbox.com

DomT4 commented:
> bugs.debian.org

Done.

> *.ietf.org

Done, although the number of impacted formulae is limited here, and I can't imagine it expanding *that* much.

> dl.google.com `

Affects 4 formulae. I'm not sure 4 formulae deserve an audit check, to be honest. For that limited number, I'd rather we just correct folks manually with the next update.

> *.googlecode.com

Already implemented:

```
    # Check for Google Code download urls, https:// is preferred
    urls.grep(%r[^http://.*\.googlecode\.com/files.*]) do |u|
      problem "Use https:// URLs for downloads from Google Code (url is #{u})."
    end
```

> s3.amazonaws.com

Don't really want to introduce this just yet. There's way too many custom domains which aren't covered with Amazon's primary cert, and then you get failures. See https://assets.heroku.com.s3.amazonaws.com/heroku-client/heroku-client-3.21.4.tgz for example.

> *.github.io

I'd like the maintainers to comment on this one. I worry a lot about the ` github.io ` pages where people suck at not mixing content and breaking the SSL/TLS. I'd like to manually verify all of the URLs we currently use for ` github.io ` can be switched to SSL/TLS by default without breaking people's stuff first.

> *.python.org

I can only find like... four examples where we're not already using ` https:// ` links here. I don't think the trend is going in the opposite direction, so I'm hesitant to include it.

> *.dropbox.com

Can only find one example inside a formula. Don't really want a rule for just the one formula, personally.

DomT4 commented:
Changes in recent commit:

* Added savannah. No auto-redirect. Sadly, their SSL/TLS only covers the homepage links, not the URLs yet. There's a note of this in the audit documentation.
* Added bugs.debian.org. It's not an overly common one yet, but given we nudge people towards Debian in places, it seemed like it could grow. Only applies to patches, as I couldn't find much usage of it elsewhere.
* Added ietf.org.
* Changed the word ` links ` to ` urls ` in the url audit section, to clarify the failure is on the url and not automatically the homepages as well.

Nothing huge changed.

vszakats commented:
> Don't really want to introduce this just yet. There's way too many custom domains which aren't covered > with Amazon's primary cert, and then you get failures. See 
> https://assets.heroku.com.s3.amazonaws.com/heroku-client/heroku-client-3.21.4.tgz for example.

This is solvable by requiring the format `https://s3.amazonaws.com/my-bucket-name/path/file` as opposed to `https://my-bucket-name.s3.amazonaws.com/path/file`, where the latter indeed may fail if `my-bucket-name` contains a dot. The former method should always work though.

This is the correct URL for your example:
https://s3.amazonaws.com/assets.heroku.com/heroku-client/heroku-client-3.21.4.tgz

> I worry a lot about the github.io pages where people suck at not mixing content

url/head/mirror is always downloading a single file, so mixed-content is never an issue there.

I removed dl.dropbox.com from my comment, it's indeed a single instance, so I made a pull for that formula instead, problem solved.

Thanks for the rest.

vszakats commented:
On S3 bucket URLs, here's the original docs:
https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html#access-bucket-intro

DomT4 commented:
I'm not particularly high on actually changing the upstream link though. If upstream are offering, for example ` http://assets.heroku.com.s3.amazonaws.com/heroku-client/heroku-client-3.21.4.tgz `, and we change that to ` https://s3.amazonaws.com/heroku-client/heroku-client-3.21.4.tgz `, It's going to confuse people. 

I don't mind shuffling folks between HTTPS links and HTTP, but I'm not overly high on actually changing those links from the stated upstream ones. If Heroku and co have actually changed their links to reference the ` https://s3.amazonaws.com/heroku-client/heroku-client-3.21.4.tgz ` situation or similar, I'll change more happily.

> url/head/mirror is always downloading a single file, so mixed-content is never an issue there.

Fair point. Still, examples seem limited to 3:

```
jq.rb
libyubikey.rb
ykclient.rb
```

We can change those manually without a rule, IMO.

vszakats commented:
You could enforce https without rewriting upstream links for those that are known to always work.
F.e. this regex:
`^http://(?:[A-Za-z0-9\-_]+\.|)s3(?:\-[a-zA-Z-]+|)\.amazonaws\.com/`
would catch pure `http://s3.amazonaws.com/` domains, plus ones with a compliant (dot-free) subdomain, including the optional regions specifiers.

[ For the rest, they are in this very pending commit, so I agree :) ]

vszakats commented:
Another regexp to catch S3 urls that make https fail due to the dotted subdomain:
`^http://[\w\-]+\.[\w\-\.]+\.s3(?:-[\w-]+|)\.amazonaws\.com/`

It may be useful to display a warning / suggestion to update the link to https-compliant form.

DomT4 commented:
It looks like ` http://s3.amazonaws.com/ ` is already pretty limited:

```
cfengine.rb:  url 'http://s3.amazonaws.com/cfengine.package-repos/tarballs/cfengine-3.6.3.tar.gz'
cfengine.rb:    url "http://s3.amazonaws.com/cfengine.package-repos/tarballs/masterfiles-3.6.3.tar.gz"
passenger.rb:  url 'http://s3.amazonaws.com/phusion-passenger/releases/passenger-4.0.57.tar.gz'
```

In regards to:
> It may be useful to display a warning / suggestion to update the link to https-compliant form.

I don't know if it's possible for the audit to throw non-fatal errors, i.e. errors that won't cause the ` brew audit ` stage to fail on the bot. I'm not sure we want people trying to just do simple updates to run into big red crosses next to their changes, it tends to scare people away, heh. I don't know. I'm hesitant on the S3 stuff, given the janky mix of domains. I'm going to leave that particular call up to the maintainers.

If the maintainers are happy without the S3 stuff for now, I'm content with the state of these changes at this point in time.

DomT4 commented:
New commit fixes a typo I made. Otherwise, no new changes.
